### DISCUSSION TITLE
br is a void element tag and must neither have `children` nor use `dangerouslySetInnerHTML`.

### DISCUSSION THREAD

**Discussion by vedraj360** [remarkjs/remark#541] [Q&A] [answered]
I am fetching readme.md file from git repo and when I get response I add it react markdown this error shows up.

this is my error 

Error: br is a void element tag and must neither have `children` nor use `dangerouslySetInnerHTML`.
    in br (created by VirtualHtml)
    in VirtualHtml (created by ReactMarkdown)
    in p (created by ReactMarkdown)
    in div (created by Root)
    in Root (created by ReactMarkdown)

The repo I am fetching is https://github.com/Chrisvin/BlobBackgroundLayout/blob/master/README.md

This is my code.

```
  useEffect(() => {
    if (!loading) {
      let git = gitInfo(gitLink);
      getGitInfo(git.username, git.repositoryname)
        .then((res) => {
          if (res != null) {
            setGitResponse(res.data);
            // console.log(res.data);
          }
        })
        .catch((err) => {
          console.log(err);
        });
    }
    
  }, [loading, gitLink]);

 <ReactMarkdown
          className="projectDetail__markdown"
          escapeHtml={false}
          transformImageUri={(uri) => checkURI(uri, gitLink)}
          source={markDown}
          // astplugins={[toc]}
          renderers={{ code: CodeBlock, table: TableWrapper }}
          // plugins={[toc]}
        />

/*
codeblock is like code wrapper it detects the code backticks and table wrapper is used to wrap the rendered table into the div block.

/*
 ** Check image from the markdown text is correct or not or convert according to requirement.
 @param uri
 ? return formated uri

 */
const checkURI = (uri, gitLink) => {
  if (
    uri.includes("shields.io") ||
    uri.includes("jitpack.io") ||
    uri.includes("travis-ci.org") ||
    uri.includes("codecov.io") ||
    uri.includes("api.bintray.com") ||
    uri.includes("api.codacy.com")
  ) {
    return uri;
  } else if (uri.startsWith("/") || uri.startsWith("./")) {
    let x = uri.split(".com");
    let y = `https://raw.githubusercontent.com` + x[1];
    let z = y.replace("/blob", "");
    return z;
  } else if (uri.includes("https://raw.githubusercontent.com")) {
    return uri;
  } else if (!uri.includes("http") || !uri.includes("https")) {
    let split = gitLink.split("/");
    let userUri = "/" + split[3] + "/" + split[4] + "/master/" + uri;
    let y = `https://raw.githubusercontent.com/` + userUri;
    return y;
  } else if (uri.startsWith("https://github.com/")) {
    let x = uri.split(".com");
    // console.log(x);
    let y = `https://raw.githubusercontent.com` + x[1];
    let z = y.replace("/blob", "");
    return z;
  }
  return uri;
};

```

**Comment by wooorm** [comment, marked as answer]
Please value our time: https://github.com/remarkjs/.github/blob/main/support.md

**Reply by vedraj360** [reply]
I value your time but I don't know what to do now. I have spend 2 hrs trying to solve this error but not able to solve. I am fetching readme.md file from git repo and when I get response I add it react markdown this error shows up.

**Reply by wooorm** [reply]
That is great to hear! However, the link I posted importantly mentions: **Spend time framing a question**. But this isn’t really a question: it’s only a stack trace! With such a broad “question”, I have nothing to go on. I don’t know about your code. It seems like a problem with React? Why ask it here? It could be anything!

What is going to happen now, is that probably we’re entering a loop of me asking “why”, “what”, “how”, etc, you giving some more information, and that repeats 5 times, until I have enough to go on. Could you try and give some more information up front?

**Reply by vedraj360** [reply]
I have update my question .

**Reply by wooorm** [reply]
Could you use fenced code blocks for your code? See: https://commonmark.org/help/. It is very hard for me to read right now. Thanks!

Example:

~~~markdown
```
# code block
print '3 backticks or'
print 'indent 4 spaces'
```
~~~

**Reply by vedraj360** [reply]
want to tell you that if you paste markdown text of this repo (https://raw.githubusercontent.com/Chrisvin/BlobBackgroundLayout/master/README.md)  on the demo page of react markdown https://remarkjs.github.io/react-markdown/ this demo page is also crashing. 
Thanks for helping me

**Reply by wooorm** [reply]
Couple of things:

- `escapeHtml` is a security vulnerability, I sure hope you know what you’re doing!
- What is `checkURI`?
- What is `markDown`?
- What are your renderers?

The markdown you are loading is invalid HTML. It includes `</br>` tags (which doesn’t make sense).
See the notes on HTML in the readme of react-markdown: https://github.com/remarkjs/react-markdown#appendix-a-html-in-markdown
But still: parsing HTML isn’t suggested.

Maybe, could you take couple of steps back? What are you trying to achieve?
See [rubber ducking](https://rubberduckdebugging.com/) and [xy problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem/66378#66378)

**Reply by vedraj360** [reply]
```</br>(OR)
</br>Download the latest demo apk from [releases](https://github.com/Chrisvin/BlobBackgroundLayout/releases).
```
this particular tag is causing problem in rendering I think

**Reply by vedraj360** [reply]
</br>(OR)
</br>Download the latest demo apk from [releases](https://github.com/Chrisvin/BlobBackgroundLayout/releases).

this particular tag is causing problem in rendering I think.
Try to paste it in demo page.

**Reply by wooorm** [reply]
Could you check the rest of my comment as well? E.g., using with-html?

**Reply by vedraj360** [reply]
when using ReactMarkdownWithHtml error is  Cannot read property 'props' of undefined at react-markdown/lib/ast-to-react.js:240

**Reply by wooorm** [reply]
That might be https://github.com/remarkjs/react-markdown/issues/385

**Reply by vedraj360** [reply]
So, I have to use  
```
const [reactContent, setMarkdownSource] = useRemark({
  remarkPlugins: [remarkMath],
  rehypePlugins: [rehypeKatex],
});
instead of 
<ReactMarkdown> </ReactMarkdown>

```

**Reply by wooorm** [reply]
If you want to properly parse HTML, that sounds nice yeah! ReactMarkdown isn’t a proper HTML parser

**Reply by vedraj360** [reply]
Thanks for your time again and help. It was nice to talk to you. +repo

**Reply by wooorm** [reply]
Awesome @vedraj360, good luck in your open source / development career! Have fun!

**Reply by ChristianMurphy** [reply]
> So, I have to use
>
> ```jsx
> const [reactContent, setMarkdownSource] = useRemark({
>   remarkPlugins: [remarkMath],
>   rehypePlugins: [rehypeKatex],
> });
> // instead of 
> <ReactMarkdown> </ReactMarkdown>
> ```

assuming you are referring to https://github.com/remarkjs/react-markdown/pull/428 or https://github.com/ChristianMurphy/react-remark
no, both still offer a component API, but *additionally* offer a hooks API if you prefer.
